core/corehttp/gateway.go

```diff
@@ -12,8 +12,7 @@ func GatewayOption(writable bool) ServeOption {
 		if err != nil {
 			return err
 		}
-		mux.Handle("/ipfs/", gateway)
-		mux.Handle("/ipns/", gateway)
+		mux.Handle("/", gateway)
```

Comment:
It may be doable by having two muxers:
- one mounted on the server at `"/"`. performs the check, possibly rewriting the path, and then calls the other:
- one which handles everything else.